File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D05.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y15D05(input: String) : Day {
    private val strings = input.lines()

    // Create Regex patterns for below functions
    private val vowels = Regex("[aeiou]")
    private val repeatedLetter = Regex("""([a-z])\1""")
    private val forbidden = Regex("ab|cd|pq|xy")
    private val repeatedDuo = Regex("""([a-z]{2}).*\1""")
    private val repeated1Between = Regex("""([a-z]).\1""")

    // Functions evaluate specific criteria
    private fun String.atLeast3Vowels() = vowels.findAll(this).count() >= 3
    private fun String.atLeastOneTwice() = repeatedLetter.containsMatchIn(this)
    private fun String.noForbiddenStrings() = !forbidden.containsMatchIn(this)
    private fun String.atLeastTwoTwice() = repeatedDuo.containsMatchIn(this)
    private fun String.repeatsWithOneBetween() = repeated1Between.containsMatchIn(this)

    private fun String.isNicePart1() = atLeast3Vowels() && atLeastOneTwice() && noForbiddenStrings()
    private fun String.isNicePart2() = atLeastTwoTwice() && repeatsWithOneBetween()

    override fun part1() = strings.count { it.isNicePart1() }

    override fun part2() = strings.count { it.isNicePart2() }
}

fun main() = Day.runDay(Y15D05::class)

//    Class creation: 31ms
//    Part 1: 255 (8ms)
//    Part 2: 55 (7ms)
//    Total time: 47ms

File: rust/advent/src/bin/2015/y15d05.rs
use fancy_regex::Regex;
use advent::utilities::get_input::get_input;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};
use std::sync::LazyLock;

type Input<'a> = &'a str;
type Output = usize;

static VOWELS: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"[aeiou]").unwrap()
});
static REPEATED_LETTER: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"([a-z])\1").unwrap()
});
static FORBIDDEN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"ab|cd|pq|xy").unwrap()
});
static REPEATED_DUO: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"([a-z]{2}).*\1").unwrap()
});
static REPEATED_1_BETWEEN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"([a-z]).\1").unwrap()
});

trait MaybeNice {
    fn at_least_three_vowels(&self) -> bool;
    fn at_least_one_twice(&self) -> bool;
    fn no_forbidden_strings(&self) -> bool;
    fn at_least_two_twice(&self) -> bool;
    fn repeats_with_one_between(&self) -> bool;
    fn is_kinda_nice(&self) -> bool;
    fn is_really_nice(&self) -> bool;
}

impl<'a> MaybeNice for &'a str {
    fn at_least_three_vowels(&self) -> bool {
        VOWELS.find_iter(self).count() >= 3
    }

    fn at_least_one_twice(&self) -> bool {
        REPEATED_LETTER.is_match(self).unwrap_or_default()
    }

    fn no_forbidden_strings(&self) -> bool {
        !FORBIDDEN.is_match(self).unwrap_or_default()
    }

    fn at_least_two_twice(&self) -> bool {
        REPEATED_DUO.is_match(self).unwrap_or_default()
    }

    fn repeats_with_one_between(&self) -> bool {
        REPEATED_1_BETWEEN.is_match(self).unwrap_or_default()
    }

    fn is_kinda_nice(&self) -> bool {
        self.at_least_three_vowels() && self.at_least_one_twice() && self.no_forbidden_strings()
    }

    fn is_really_nice(&self) -> bool {
        self.at_least_two_twice() && self.repeats_with_one_between()
    }
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 5).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn part1(strings: Input) -> Output {
    strings.lines()
        .filter(|line| line.is_kinda_nice())
        .count()
}

fn part2(strings: Input) -> Output {
    strings.lines()
        .filter(|line| line.is_really_nice())
        .count()
}

#[test]
fn default() {
    let input = get_input(15, 5).unwrap();
    assert_eq!(255, part1(&input));
    assert_eq!(55, part2(&input));
}

// Input parsed (27μs)
// 1. 255 (724μs)
// 2. 55 (2ms)
// Total: 3ms
